File: src/main/java/pl/com/tegess/controller/application/readcontroller/ApplicationReadController.java
package pl.com.tegess.controller.application.readcontroller;

import org.apache.commons.lang3.tuple.Pair;
import org.bson.types.ObjectId;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;
import org.springframework.web.bind.annotation.*;
import pl.com.tegess.controller.application.request.ApplicationData;
import pl.com.tegess.controller.application.request.LoginNumber;
import pl.com.tegess.controller.application.request.MinuteStats;
import pl.com.tegess.domain.application.ApplicationRepository;
import pl.com.tegess.domain.events.LoginService;

import java.util.Date;
import java.util.List;
import java.util.stream.Collectors;

@RestController
@Component
public class ApplicationReadController {

    @Autowired
    ApplicationRepository repository;
    @Autowired
    LoginService loginService;

    @RequestMapping(value = "/api/applications", method = RequestMethod.GET)
    public List<ApplicationData> getApplications() {
        return repository.find()
                .asList()
                .stream()
                .map(application -> new ApplicationData(
                        application.getAppId().toString(),
                        application.getFacebookAppId(),
                        application.getSecret(),
                        application.getRedirectURI()))
                .collect(Collectors.toList());

    }

    @RequestMapping(value = "/api/applications/{appId}/loginNumber", method = RequestMethod.GET)
    public LoginNumber getLoginNumber(@PathVariable String appId, @RequestParam Long from, @RequestParam Long to) {
        return new LoginNumber(loginService.countLoginForAppInPeriod(new ObjectId(appId),
                new Date(from), new Date(to)));
    }

    @RequestMapping(value = "/api/applications/{appId}/hourLoginStats", method = RequestMethod.GET)
    public List<MinuteStats> getHourLoginStats(@PathVariable String appId) {
        return loginService.getHourLoginStats(new ObjectId(appId)).stream()
                .map(pair -> new MinuteStats(pair.getKey(), pair.getValue()))
                .collect(Collectors.toList());
    }
}

File: src/main/scala/com/tegess/controller/application/ApplicationReadController.scala
package com.tegess.controller.application

import com.tegess.config.newsecurity.TokenService
import com.tegess.controller.application.ApplicationReadController.{ApplicationData, MinuteStatistics}
import com.tegess.domain.admin.Admin
import com.tegess.persistance.service.application.ApplicationService
import com.tegess.persistance.service.event.LoginEventService
import org.bson.types.ObjectId
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.security.core.Authentication
import org.springframework.stereotype.Component
import org.springframework.web.bind.annotation
import org.springframework.web.bind.annotation.{PathVariable, RequestMapping, RequestMethod}

import scala.util.{Failure, Success, Try}

@annotation.RestController
@Component
class ApplicationReadController {

  @Autowired var applicationService: ApplicationService = _
  @Autowired var loginEventService: LoginEventService = _

  @RequestMapping(value=Array("/api/applications"), method=Array(RequestMethod.GET))
  def getApplications(authentication: Authentication): List[ApplicationData] = {
    val admin = authentication.getPrincipal.asInstanceOf[Admin]
    val applications = applicationService.findAll(admin)
    for {
      application <- applications
    } yield {
      ApplicationData(application.id.toHexString,
        application.secret,
        application.fbLogin.map(_.id),
        application.redirectURL,
        application.fbLogin.map(_.secret))
    }

  }

  @RequestMapping(value=Array("/api/applications/{appId}/hourLoginStats"), method=Array(RequestMethod.GET))
  def getHourLoginStats(@PathVariable appId: String): List[MinuteStatistics] = {
    val result = for {
      applicationId <- Try(new ObjectId(appId))
    } yield loginEventService.getHourLoginStats(applicationId).map(pair => MinuteStatistics(pair._1, pair._2))

    result match {
      case Success(r) => r.toList
      case Failure(e) => throw new IllegalArgumentException("Something went wrong -", e)
    }
  }

}
object ApplicationReadController {
  case class ApplicationData(appId: String,
                             secret: String,
                             facebookAppId: Option[String],
                             facebookRedirectURI: Option[String],
                             facebookSecret: Option[String])
  case class MinuteStatistics(time: Long, amount: Long)
}
